platform/mbed_retarget.cpp

```diff
@@ -110,20 +118,138 @@ void remove_filehandle(FileHandle *file) {
 #if DEVICE_SERIAL
 extern int stdio_uart_inited;
 extern serial_t stdio_uart;
-#if MBED_CONF_PLATFORM_STDIO_CONVERT_NEWLINES
-static char stdio_in_prev;
-static char stdio_out_prev;
-#endif
 #endif
 
-static void init_serial() {
-#if DEVICE_SERIAL
+/* Private FileHandle to implement backwards-compatible functionality of
+ * direct HAL serial access for default stdin/stdout/stderr.
+ * This is not a particularly well-behaved FileHandle for a stream, which
+ * is why it's not public. People should be using UARTSerial.
+ */
+class DirectSerial : public FileHandle {
```

Comment:
Peripheral question regarding this comment - what's the difference between `UARTSerial` and `Serial`? I've been using `Serial`. Are there times when I (an Mbed OS user) should be using `UARTSerial`?

I'm not finding any documentation about `UARTSerial`.

Reply:
@kjbracey-arm @sg- ^^^

Reply:
`UARTSerial` is buffered. (This was more obvious when it was called `BufferedSerial` - name was changed to avoid collision with its out-of-tree predecessor inside the `ATParser` used by external drivers. We may rename it back).

`Serial` is unbuffered using blocking HAL calls so eats a lot of CPU time spinning waiting for serial output, and can't really be used reliably for input (it needs interrupt-speed response on read calls, but its high-level API can't be used from interrupt).

I would use `RawSerial` (direct to HAL) for synchronous CPU-blocking trace or interrupt-based IO,  or `UARTSerial` (buffered) for normal thread-based/application use. I would avoid `Serial`.

Handbook updates for `UARTSerial` were in progress, but I think they got stuck at review somewhere in November.

Then there's this: #5655. Discussions ongoing for what to do, and when - latest thought was:

 * `RawSerial` becomes a `FileHandle`, eliminating need for that internal `DirectSerial`
 * `Serial` gets dropped/deprecated
 * `UARTSerial` gets renamed as `BufferedSerial`
 * Ponder how much we really want `X.printf` methods rather than using `fprintf(X)`.